File: backend/src/services/shopifyService.ts
import crypto from "crypto";

export interface ShopifyCredentials {
  shopDomain: string;
  accessToken: string;
  apiSecretKey?: string;
}

export interface ShopifyShopInfo {
  id: number;
  name: string;
  email: string;
  domain: string;
  currency: string;
  country_name: string;
}

export interface ShopifyProduct {
  id: number;
  title: string;
  body_html?: string;
  vendor?: string;
  product_type?: string;
  created_at?: string;
  images?: Array<{ src: string }>;
  variants?: Array<{ price: string; inventory_quantity: number }>;
}

export class ShopifyService {
  private cleanDomain(domain: string): string {
    let clean = domain.trim().toLowerCase();
    clean = clean.replace(/^https?:\/\//, "");
    clean = clean.replace(/\/.*$/, "");
    if (!clean.includes(".myshopify.com") && !clean.includes(".")) {
      clean = `${clean}.myshopify.com`;
    }
    return clean;
  }

  async verifyCredentials(credentials: ShopifyCredentials): Promise<{ success: boolean; shop?: ShopifyShopInfo; error?: string }> {
    const shopDomain = this.cleanDomain(credentials.shopDomain);
    const url = `https://${shopDomain}/admin/api/2024-01/shop.json`;

    try {
      const response = await fetch(url, {
        method: "GET",
        headers: {
          "X-Shopify-Access-Token": credentials.accessToken,
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        if (response.status === 401) {
          return { success: false, error: "Invalid Access Token or unauthorized access." };
        }
        if (response.status === 404) {
          return { success: false, error: "Shop domain not found. Check your .myshopify.com URL." };
        }
        return { success: false, error: `Shopify API returned status ${response.status}` };
      }

      const data = await response.json() as { shop: ShopifyShopInfo };
      return {
        success: true,
        shop: data.shop,
      };
    } catch (err: any) {
      return {
        success: false,
        error: err.message || "Failed to reach Shopify server. Please check network connection and domain.",
      };
    }
  }

  async fetchProducts(credentials: ShopifyCredentials, limit = 50): Promise<{ success: boolean; products?: ShopifyProduct[]; count?: number; error?: string }> {
    const shopDomain = this.cleanDomain(credentials.shopDomain);
    const url = `https://${shopDomain}/admin/api/2024-01/products.json?limit=${limit}`;

    try {
      const response = await fetch(url, {
        method: "GET",
        headers: {
          "X-Shopify-Access-Token": credentials.accessToken,
          "Content-Type": "application/json",
        },
      });

      if (!response.ok) {
        return { success: false, error: `Shopify returned error ${response.status}` };
      }

      const data = await response.json() as { products: ShopifyProduct[] };
      return {
        success: true,
        products: data.products || [],
        count: (data.products || []).length,
      };
    } catch (err: any) {
      return {
        success: false,
        error: err.message || "Failed to fetch products from Shopify.",
      };
    }
  }

  async createProduct(
    credentials: ShopifyCredentials,
    productData: {
      title: string;
      body_html: string;
      vendor?: string;
      product_type?: string;
      tags?: string;
      price?: string;
      inventory_quantity?: number;
      images?: Array<{ src: string }>;
    }
  ): Promise<{ success: boolean; product?: ShopifyProduct; error?: string }> {
    const shopDomain = this.cleanDomain(credentials.shopDomain);
    const url = `https://${shopDomain}/admin/api/2024-01/products.json`;

    const payload = {
      product: {
        title: productData.title,
        body_html: productData.body_html,
        vendor: productData.vendor || "Local Baba Seller",
        product_type: productData.product_type || "General",
        tags: productData.tags || "",
        variants: [
          {
            price: productData.price || "0.00",
            inventory_quantity: productData.inventory_quantity ?? 100,
          },
        ],
        images: productData.images || [],
      },
    };

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "X-Shopify-Access-Token": credentials.accessToken,
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        if (response.status === 401) {
          return {
            success: false,
            error:
              "Invalid Admin API Access Token. Please go to Integrations and re-enter a valid token from your Shopify Admin (Develop Apps -> API Credentials).",
          };
        }
        if (response.status === 403) {
          return {
            success: false,
            error:
              "Access denied. Your token does not have 'write_products' scope. Go to Shopify Admin -> Develop Apps -> Edit Scopes and enable write_products.",
          };
        }
        const errorText = await response.text();
        return { success: false, error: `Shopify returned error ${response.status}: ${errorText}` };
      }

      const data = (await response.json()) as { product: ShopifyProduct };
      return {
        success: true,
        product: data.product,
      };
    } catch (err: any) {
      return {
        success: false,
        error: err.message || "Failed to create product on Shopify.",
      };
    }
  }

  verifyHmac(rawBody: string, hmacHeader: string, secretKey: string): boolean {
    if (!hmacHeader || !secretKey) return false;
    const generatedHmac = crypto
      .createHmac("sha256", secretKey)
      .update(rawBody, "utf8")
      .digest("base64");
    return crypto.timingSafeEqual(Buffer.from(generatedHmac), Buffer.from(hmacHeader));
  }
}

export const shopifyService = new ShopifyService();

File: backend/app/services/shopify_service.py
import base64
import hashlib
import hmac
import re

import requests

API_VERSION = "2024-01"
TIMEOUT = 15


def clean_domain(domain: str) -> str:
    clean = domain.strip().lower()
    clean = re.sub(r"^https?://", "", clean)
    clean = re.sub(r"/.*$", "", clean)
    if ".myshopify.com" not in clean and "." not in clean:
        clean = f"{clean}.myshopify.com"
    return clean


def _headers(access_token: str) -> dict:
    return {"X-Shopify-Access-Token": access_token, "Content-Type": "application/json"}


def verify_credentials(shop_domain: str, access_token: str) -> dict:
    domain = clean_domain(shop_domain)
    try:
        res = requests.get(
            f"https://{domain}/admin/api/{API_VERSION}/shop.json",
            headers=_headers(access_token),
            timeout=TIMEOUT,
        )
    except requests.RequestException as exc:
        return {"success": False, "error": f"Could not reach Shopify: {exc}"}

    if res.status_code == 401:
        return {"success": False, "error": "Invalid Shopify Admin API access token"}
    if res.status_code == 404:
        return {"success": False, "error": "Shop not found — check the store domain"}
    if not res.ok:
        return {"success": False, "error": f"Shopify verification failed ({res.status_code})"}

    shop = res.json().get("shop", {})
    return {"success": True, "shop": shop}


def fetch_products(shop_domain: str, access_token: str, limit: int = 50) -> dict:
    domain = clean_domain(shop_domain)
    try:
        res = requests.get(
            f"https://{domain}/admin/api/{API_VERSION}/products.json",
            params={"limit": limit},
            headers=_headers(access_token),
            timeout=TIMEOUT,
        )
    except requests.RequestException as exc:
        return {"success": False, "error": f"Could not reach Shopify: {exc}"}

    if not res.ok:
        return {"success": False, "error": f"Failed to fetch products ({res.status_code})"}

    return {"success": True, "products": res.json().get("products", [])}


def create_product(shop_domain: str, access_token: str, product_data: dict) -> dict:
    domain = clean_domain(shop_domain)
    payload = {
        "product": {
            "title": product_data.get("title"),
            "body_html": product_data.get("body_html", ""),
            "vendor": product_data.get("vendor"),
            "product_type": product_data.get("product_type"),
            "tags": product_data.get("tags"),
            "variants": [
                {
                    "price": product_data.get("price", "0.00"),
                    "inventory_quantity": product_data.get("inventory_quantity", 0),
                }
            ],
            "images": product_data.get("images", []),
        }
    }

    try:
        res = requests.post(
            f"https://{domain}/admin/api/{API_VERSION}/products.json",
            json=payload,
            headers=_headers(access_token),
            timeout=TIMEOUT,
        )
    except requests.RequestException as exc:
        return {"success": False, "error": f"Could not reach Shopify: {exc}"}

    if res.status_code == 401:
        return {"success": False, "error": "Invalid Shopify Admin API access token"}
    if res.status_code == 403:
        return {
            "success": False,
            "error": "This access token is missing the write_products scope",
        }
    if not res.ok:
        return {"success": False, "error": f"Failed to create product ({res.status_code}): {res.text[:300]}"}

    return {"success": True, "product": res.json().get("product", {})}


def verify_hmac(raw_body: bytes, hmac_header: str | None, secret_key: str) -> bool:
    if not hmac_header or not secret_key:
        return False
    digest = hmac.new(secret_key.encode("utf-8"), raw_body, hashlib.sha256).digest()
    computed = base64.b64encode(digest).decode("utf-8")
    return hmac.compare_digest(computed, hmac_header)
